h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java

```diff
@@ -522,76 +521,137 @@ private static int initalizeFromChunkIds(Frame f, int[] chunks, Vec.Reader[] vec
 
         // extract predictors
         int nonZeroCount = 0;
-        int currentRow = 0;
+        int rowPointer = 0;
         int currentCol = 0;
         int rwRow = 0;
+        
+        int rowHeaderRowPointer = 0;
+        int rowHeaderColPointer = 0;
+        
+        int lastNonZeroRow = 0;
 
         for (Integer chunk : chunks) {
             for(long i = f.anyVec().espc()[chunk]; i < f.anyVec().espc()[chunk+1]; i++) {
                 if (w != null && w.at(i) == 0) continue;
-
-                final int startNonZeroCount = nonZeroCount;
-                // enlarge final data arrays by 2x if needed
+                actualRows++;
+                if(rowHeaderColPointer == SPARSE_MATRIX_DIM){
+                    rowHeaderColPointer = 0;
+                    rowHeaderRowPointer++;
+                }
+                boolean foundNonZero = false;
 
                 for (int j = 0; j < di._cats; ++j) {
-                    data[currentRow][currentCol] = 1; //one-hot encoding
+                    if(currentCol == SPARSE_MATRIX_DIM){
+                        currentCol = 0;
+                        rowPointer++;
+                    }
+                    data[rowPointer][currentCol] = 1; //one-hot encoding
+                    
+                    if(!foundNonZero){
+                        foundNonZero = true;
+                        for (int k = lastNonZeroRow; k < actualRows; k++) {
+                        rowHeaders[rowHeaderRowPointer][rowHeaderColPointer++] = nonZeroCount;
+                        }
+                        lastNonZeroRow = actualRows;
+                    }
                     if (vecs[j].isNA(i)) {
-                        colIndex[currentRow][currentCol++] = di.getCategoricalId(j, Float.NaN);
+                        colIndex[rowPointer][currentCol++] = di.getCategoricalId(j, Float.NaN);
                     } else {
-                        colIndex[currentRow][currentCol++] = di.getCategoricalId(j, vecs[j].at8(i));
+                        colIndex[rowPointer][currentCol++] = di.getCategoricalId(j, vecs[j].at8(i));
                     }
                     nonZeroCount++;
                 }
 
                 for (int j = 0; j < di._nums; ++j) {
+                    if(currentCol == SPARSE_MATRIX_DIM){
+                        currentCol = 0;
```

Comment:
I do believe a single check at each round will in the end (after all optimizations kick in) be a little bit faster than calculating the indices each time.